File: iOS/smap/ViewModels/GroupViewModel.swift
import Foundation
import SwiftUI
import Combine

class GroupViewModel: ObservableObject {
    @Published var groups: [SmapGroup] = []
    @Published var selectedGroup: SmapGroup?
    @Published var groupMembers: [SmapGroupMember] = []
    @Published var groupStats: GroupStats?
    @Published var totalMembers: Int = 0
    
    // Loading States
    @Published var isLoading: Bool = false
    @Published var isCreating: Bool = false
    @Published var isUpdating: Bool = false
    @Published var isDeleting: Bool = false
    @Published var isJoining: Bool = false
    
    // Error States
    @Published var errorMessage: String?
    @Published var showError: Bool = false
    
    private let groupService = GroupService.shared
    
    // MARK: - Group Management
    
    /// 내 그룹 목록 조회
    func fetchGroups() {
        isLoading = true
        Task {
            do {
                let fetchedGroups = try await groupService.getCurrentUserGroups()
                DispatchQueue.main.async {
                    self.groups = fetchedGroups
                    self.isLoading = false
                    
                    if let selected = self.selectedGroup,
                       let updated = fetchedGroups.first(where: { $0.sgt_idx == selected.sgt_idx }) {
                        self.selectedGroup = updated
                    } else if self.selectedGroup == nil && !fetchedGroups.isEmpty {
                        // 선택된 그룹이 없으면 첫번째 그룹 선택 (옵션)
                        // self.selectedGroup = fetchedGroups.first
                    }
                    
                    // 요약 정보도 갱신
                    self.fetchGroupSummary()
                }
                }
            } catch {
                DispatchQueue.main.async {
                    self.handleError(error)
                    self.isLoading = false
                }
            }
        }
    }
    
    /// 그룹 생성
    func createGroup(title: String, memo: String) {
        guard !title.isEmpty else { return }
        isCreating = true
        Task {
            do {
                let newGroup = try await groupService.createGroup(title: title, memo: memo)
                DispatchQueue.main.async {
                    self.groups.append(newGroup)
                    self.selectedGroup = newGroup
                    self.isCreating = false
                    self.fetchGroups() // 목록 갱신
                    // Notify other views that groups have changed
                    NotificationCenter.default.post(name: NSNotification.Name("groupsDidChange"), object: nil)
                }
            } catch {
                DispatchQueue.main.async {
                    self.handleError(error)
                    self.isCreating = false
                }
            }
        }
    }
    
    /// 그룹 수정
    func updateGroup(sgtIdx: Int, title: String, memo: String) {
        guard !title.isEmpty else { return }
        isUpdating = true
        Task {
            do {
                // 수정된 그룹 정보 반환을 가정하거나, 다시 조회
                let updatedGroup = try await groupService.updateGroup(sgtIdx: sgtIdx, title: title, memo: memo)
                DispatchQueue.main.async {
                    if let index = self.groups.firstIndex(where: { $0.sgt_idx == sgtIdx }) {
                        self.groups[index] = updatedGroup
                    }
                    if self.selectedGroup?.sgt_idx == sgtIdx {
                        self.selectedGroup = updatedGroup
                    }
                    self.isUpdating = false
                }
            } catch {
                DispatchQueue.main.async {
                    self.handleError(error)
                    self.isUpdating = false
                }
            }
        }
    }
    
    /// 그룹 삭제 (소프트 삭제)
    func deleteGroup(sgtIdx: Int) {
        isDeleting = true
        Task {
            do {
                let success = try await groupService.deleteGroup(sgtIdx: sgtIdx)
                if success {
                    DispatchQueue.main.async {
                        self.groups.removeAll(where: { $0.sgt_idx == sgtIdx })
                        if self.selectedGroup?.sgt_idx == sgtIdx {
                            self.selectedGroup = nil
                            self.groupMembers = []
                            self.groupStats = nil
                        }
                        self.isDeleting = false
                        // Notify other views that groups have changed
                        NotificationCenter.default.post(name: NSNotification.Name("groupsDidChange"), object: nil)
                    }
                }
            } catch {
                DispatchQueue.main.async {
                    self.handleError(error)
                    self.isDeleting = false
                }
            }
        }
    }
    
    /// 그룹 가입 (초대 코드)
    func joinGroup(inviteCode: String) {
        guard !inviteCode.isEmpty else { return }
        isJoining = true
        Task {
            do {
                let success = try await groupService.joinGroup(inviteCode: inviteCode)
                if success {
                    DispatchQueue.main.async {
                        self.isJoining = false
                        self.fetchGroups() // 목록 갱신 및 UI 이동
                        // Notify other views that groups have changed
                        NotificationCenter.default.post(name: NSNotification.Name("groupsDidChange"), object: nil)
                    }
                }
            } catch {
                DispatchQueue.main.async {
                    self.handleError(error)
                    self.isJoining = false
                }
            }
        }
    }
    
    // MARK: - Detail Info
    
    /// 그룹 멤버 조회
    func fetchGroupMembers(sgtIdx: Int) {
        // 이미 로딩 중이면 스킵하거나, 개별 로딩 상태 관리
        Task {
            do {
                let members = try await groupService.getGroupMembers(sgtIdx: sgtIdx)
                DispatchQueue.main.async {
                    self.groupMembers = members
                }
            } catch {
                print("❌ [GroupViewModel] 멤버 조회 실패: \(error)")
            }
        }
    }
    
    /// 그룹 통계 조회
    func fetchGroupStats(sgtIdx: Int) {
        Task {
            do {
                let stats = try await groupService.getGroupStats(sgtIdx: sgtIdx)
                DispatchQueue.main.async {
                    self.groupStats = stats
                }
            } catch {
                print("❌ [GroupViewModel] 통계 조회 실패: \(error)")
            }
        }
    }
    
    /// 멤버 권한 변경 (리더 위임 등)
    func updateMemberRole(member: SmapGroupMember, isLeader: Bool) {
        guard let group = selectedGroup else { return }
        Task {
            do {
                let success = try await groupService.updateMemberRole(sgtIdx: group.sgt_idx, mtIdx: member.mt_idx, isLeader: isLeader)
                if success {
                    self.fetchGroupMembers(sgtIdx: group.sgt_idx) // 멤버 목록 갱신
                }
            } catch {
                DispatchQueue.main.async { self.handleError(error) }
            }
        }
    }
    
    /// 멤버 강퇴
    func removeMember(member: SmapGroupMember) {
        guard let group = selectedGroup else { return }
        Task {
            do {
                let success = try await groupService.removeMember(sgtIdx: group.sgt_idx, mtIdx: member.mt_idx)
                if success {
                    self.fetchGroupMembers(sgtIdx: group.sgt_idx) // 멤버 목록 갱신
                    self.fetchGroupStats(sgtIdx: group.sgt_idx) // 통계도 갱신
                    self.fetchGroups() // 내 그룹 목록 갱신
                }
            } catch {
                DispatchQueue.main.async { self.handleError(error) }
            }
        }
    }
    
    /// 그룹 나가기 (탈퇴)
    func leaveGroup(sgtIdx: Int) {
        guard let currentUser = AuthService.shared.getUserData(),
              let mtIdx = currentUser.mt_idx else {
            self.errorMessage = "사용자 정보를 찾을 수 없습니다."
            self.showError = true
            return
        }
        
        isDeleting = true
        Task {
            do {
                let success = try await groupService.removeMember(sgtIdx: sgtIdx, mtIdx: mtIdx)
                if success {
                    DispatchQueue.main.async {
                        self.groups.removeAll(where: { $0.sgt_idx == sgtIdx })
                        if self.selectedGroup?.sgt_idx == sgtIdx {
                            self.selectedGroup = nil
                            self.groupMembers = []
                            self.groupStats = nil
                        }
                        self.isDeleting = false
                        // 목록 갱신 및 알림
                        self.fetchGroups()
                        NotificationCenter.default.post(name: NSNotification.Name("groupsDidChange"), object: nil)
                    }
                }
            } catch {
                DispatchQueue.main.async {
                    self.handleError(error)
                    self.isDeleting = false
                }
            }
        }
    }
    
    private func handleError(_ error: Error) {
        if let networkError = error as? NetworkError {
            self.errorMessage = networkError.userMessage
        } else {
            self.errorMessage = error.localizedDescription
        }
        self.showError = true
    }
    
    /// 그룹 요약 정보 조회 (총 그룹, 총 멤버)
    func fetchGroupSummary() {
        Task {
            do {
                let summary = try await groupService.getGroupSummary()
                DispatchQueue.main.async {
                    self.totalMembers = summary.total_members
                }
            } catch {
                print("❌ [GroupViewModel] 요약 정보 조회 실패: \(error)")
            }
        }
    }
}

File: Android/app/src/main/java/com/dmonster/smap/ui/group/GroupViewModel.kt
package com.dmonster.smap.ui.group

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.dmonster.smap.data.model.SmapGroup
import com.dmonster.smap.data.model.SmapGroupMember
import com.dmonster.smap.data.service.AuthService
import com.dmonster.smap.data.service.GroupService
import com.dmonster.smap.data.GlobalEvent
import com.dmonster.smap.data.GlobalEventBus
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

/**
 * 그룹 페이지 ViewModel
 */
@HiltViewModel
class GroupViewModel @Inject constructor(
    private val groupService: GroupService,
    private val authService: AuthService
) : ViewModel() {

    companion object {
        private const val TAG = "GroupViewModel"
    }

    // Current View State
    enum class ViewState { LIST, DETAIL }
    private val _currentView = MutableStateFlow(ViewState.LIST)
    val currentView: StateFlow<ViewState> = _currentView.asStateFlow()

    // Loading States
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _isMembersLoading = MutableStateFlow(false)
    val isMembersLoading: StateFlow<Boolean> = _isMembersLoading.asStateFlow()

    // Error State
    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    // Success Message
    private val _successMessage = MutableStateFlow<String?>(null)
    val successMessage: StateFlow<String?> = _successMessage.asStateFlow()

    // Data States
    private val _groups = MutableStateFlow<List<SmapGroup>>(emptyList())
    val groups: StateFlow<List<SmapGroup>> = _groups.asStateFlow()

    private val _selectedGroup = MutableStateFlow<SmapGroup?>(null)
    val selectedGroup: StateFlow<SmapGroup?> = _selectedGroup.asStateFlow()

    private val _groupMembers = MutableStateFlow<List<SmapGroupMember>>(emptyList())
    val groupMembers: StateFlow<List<SmapGroupMember>> = _groupMembers.asStateFlow()

    private val _groupMemberCounts = MutableStateFlow<Map<Int, Int>>(emptyMap())
    val groupMemberCounts: StateFlow<Map<Int, Int>> = _groupMemberCounts.asStateFlow()

    private val _groupScheduleCounts = MutableStateFlow<Map<Int, Int>>(emptyMap())
    private val _groupLocationCounts = MutableStateFlow<Map<Int, Int>>(emptyMap())
    
    val weeklyScheduleCount: StateFlow<Int> = combine(_selectedGroup, _groupScheduleCounts) { group, counts ->
        group?.let { counts[it.sgtIdx] } ?: 0
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), 0)

    val totalLocationCount: StateFlow<Int> = combine(_selectedGroup, _groupLocationCounts) { group, counts ->
        group?.let { counts[it.sgtIdx] } ?: 0
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), 0)

    // Dialog States
    private val _showCreateDialog = MutableStateFlow(false)
    val showCreateDialog: StateFlow<Boolean> = _showCreateDialog.asStateFlow()

    private val _showEditDialog = MutableStateFlow(false)
    val showEditDialog: StateFlow<Boolean> = _showEditDialog.asStateFlow()

    private val _showDeleteDialog = MutableStateFlow(false)
    val showDeleteDialog: StateFlow<Boolean> = _showDeleteDialog.asStateFlow()

    private val _showShareDialog = MutableStateFlow(false)
    val showShareDialog: StateFlow<Boolean> = _showShareDialog.asStateFlow()

    private val _showMemberManageDialog = MutableStateFlow(false)
    val showMemberManageDialog: StateFlow<Boolean> = _showMemberManageDialog.asStateFlow()

    private val _selectedMember = MutableStateFlow<SmapGroupMember?>(null)
    val selectedMember: StateFlow<SmapGroupMember?> = _selectedMember.asStateFlow()

    private val _showInviteBottomSheet = MutableStateFlow(false)
    val showInviteBottomSheet: StateFlow<Boolean> = _showInviteBottomSheet.asStateFlow()

    private val _showQRCodeDialog = MutableStateFlow(false)
    val showQRCodeDialog: StateFlow<Boolean> = _showQRCodeDialog.asStateFlow()

    // Invite Code
    private val _inviteCode = MutableStateFlow("")
    val inviteCode: StateFlow<String> = _inviteCode.asStateFlow()

    // Action Loading States
    private val _isCreating = MutableStateFlow(false)
    val isCreating: StateFlow<Boolean> = _isCreating.asStateFlow()

    private val _isJoining = MutableStateFlow(false)
    val isJoining: StateFlow<Boolean> = _isJoining.asStateFlow()

    init {
        loadGroups()
    }

    // MARK: - Data Loading

    fun loadGroups() {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null

            try {
                val groups = groupService.getGroups()
                _groups.value = groups
                Log.d(TAG, "✅ ${groups.size}개 그룹 로드")

                // Load member counts for each group
                val memberCounts = groups.associate { it.sgtIdx to (it.memberCount ?: 0) }.toMutableMap()
                _groupMemberCounts.value = memberCounts

                // Optionally refresh secondary counts if needed (but API memberCount should be reliable)
                for (group in groups) {
                    try {
                        val members = groupService.getGroupMembers(group.sgtIdx)
                        memberCounts[group.sgtIdx] = members.size
                        _groupMemberCounts.value = memberCounts.toMap()
                    } catch (e: Exception) {
                        Log.w(TAG, "⚠️ [loadGroups] Failed to get detailed members for group ${group.sgtIdx}")
                    }
                }

            } catch (e: Exception) {
                Log.e(TAG, "❌ 그룹 로드 실패", e)
                _errorMessage.value = "그룹을 불러오는데 실패했습니다"
            } finally {
                _isLoading.value = false
            }
        }
    }

    fun loadGroupMembers(group: SmapGroup) {
        viewModelScope.launch {
            _isMembersLoading.value = true

            try {
                val members = groupService.getGroupMembers(group.sgtIdx)
                _groupMembers.value = members
                Log.d(TAG, "✅ ${members.size}명 멤버 로드")

                // Update member count
                _groupMemberCounts.value = _groupMemberCounts.value.toMutableMap().apply {
                    put(group.sgtIdx, members.size)
                }

                // Load schedule count
                val scheduleCount = groupService.getGroupScheduleCount(group.sgtIdx)
                Log.d(TAG, "📊 [loadGroupMembers] Schedule Count: $scheduleCount")
                _groupScheduleCounts.value = _groupScheduleCounts.value.toMutableMap().apply {
                    put(group.sgtIdx, scheduleCount)
                }

                // Load location count
                val locationCount = groupService.getGroupLocationCount(group.sgtIdx)
                Log.d(TAG, "📊 [loadGroupMembers] Location Count: $locationCount")
                _groupLocationCounts.value = _groupLocationCounts.value.toMutableMap().apply {
                    put(group.sgtIdx, locationCount)
                }

            } catch (e: Exception) {
                Log.e(TAG, "❌ 멤버/통계 로드 실패", e)
                _errorMessage.value = "정보를 불러오는데 실패했습니다"
            } finally {
                _isMembersLoading.value = false
            }
        }
    }

    // MARK: - Navigation

    fun selectGroup(group: SmapGroup) {
        _selectedGroup.value = group
        _currentView.value = ViewState.DETAIL
        loadGroupMembers(group)
    }

    fun backToList() {
        _currentView.value = ViewState.LIST
        _selectedGroup.value = null
        _groupMembers.value = emptyList()
    }

    // MARK: - Dialog Controls

    fun showCreateDialog() { _showCreateDialog.value = true }
    fun hideCreateDialog() { _showCreateDialog.value = false }

    fun showEditDialog() { _showEditDialog.value = true }
    fun hideEditDialog() { _showEditDialog.value = false }

    fun showDeleteDialog() { _showDeleteDialog.value = true }
    fun hideDeleteDialog() { _showDeleteDialog.value = false }

    fun showShareDialog() { _showShareDialog.value = true }
    fun hideShareDialog() { _showShareDialog.value = false }

    fun showInviteBottomSheet() { _showInviteBottomSheet.value = true }
    fun hideInviteBottomSheet() { _showInviteBottomSheet.value = false }

    fun showQRCodeDialog() { _showQRCodeDialog.value = true }
    fun hideQRCodeDialog() { _showQRCodeDialog.value = false }

    fun showMemberManageDialog(member: SmapGroupMember) {
        _selectedMember.value = member
        _showMemberManageDialog.value = true
    }
    fun hideMemberManageDialog() {
        _showMemberManageDialog.value = false
        _selectedMember.value = null
    }

    // MARK: - Actions

    fun updateInviteCode(code: String) {
        // 대문자로 변환하고 영문 알파벳(A-Z)과 숫자(0-9)만 허용 (한글 제외)
        val filtered = code.uppercase().filter { it in 'A'..'Z' || it in '0'..'9' }
        _inviteCode.value = filtered
    }

    fun joinGroupByCode() {
        val code = _inviteCode.value.trim()
        if (code.isEmpty()) return

        viewModelScope.launch {
            _isJoining.value = true
            try {
                val success = groupService.joinGroupByCode(code)
                if (success) {
                    _successMessage.value = "그룹에 가입되었습니다"
                    _inviteCode.value = ""
                    loadGroups()
                    GlobalEventBus.tryEmit(GlobalEvent.GroupsChanged)
                } else {
                    _errorMessage.value = "그룹 가입에 실패했습니다"
                }
            } catch (e: Exception) {
                _errorMessage.value = "그룹 가입 중 오류가 발생했습니다"
            } finally {
                _isJoining.value = false
            }
        }
    }

    fun createGroup(title: String, description: String) {
        if (title.isBlank()) return

        viewModelScope.launch {
            _isCreating.value = true
            try {
                val group = groupService.createGroup(title, description.ifBlank { null })
                if (group != null) {
                    _successMessage.value = "그룹이 생성되었습니다"
                    hideCreateDialog()
                    loadGroups()
                    GlobalEventBus.tryEmit(GlobalEvent.GroupsChanged)
                } else {
                    _errorMessage.value = "그룹 생성에 실패했습니다"
                }
            } catch (e: Exception) {
                _errorMessage.value = "그룹 생성 중 오류가 발생했습니다"
            } finally {
                _isCreating.value = false
            }
        }
    }

    fun updateGroup(title: String, description: String) {
        val group = _selectedGroup.value ?: return
        if (title.isBlank()) return

        viewModelScope.launch {
            try {
                val success = groupService.updateGroup(group.sgtIdx, title, description.ifBlank { null })
                if (success) {
                    _successMessage.value = "그룹이 수정되었습니다"
                    hideEditDialog()
                    loadGroups()
                    // Update selected group
                    _selectedGroup.value = group.copy(
                        sgtTitle = title,
                        sgtMemo = description
                    )
                    GlobalEventBus.tryEmit(GlobalEvent.GroupsChanged)
                } else {
                    _errorMessage.value = "그룹 수정에 실패했습니다"
                }
            } catch (e: Exception) {
                _errorMessage.value = "그룹 수정 중 오류가 발생했습니다"
            }
        }
    }

    fun deleteGroup() {
        val group = _selectedGroup.value ?: return

        viewModelScope.launch {
            try {
                val success = groupService.deleteGroup(group.sgtIdx)
                if (success) {
                    _successMessage.value = "그룹이 삭제되었습니다"
                    hideDeleteDialog()
                    backToList()
                    loadGroups()
                    GlobalEventBus.tryEmit(GlobalEvent.GroupsChanged)
                } else {
                    _errorMessage.value = "그룹 삭제에 실패했습니다"
                }
            } catch (e: Exception) {
                _errorMessage.value = "그룹 삭제 중 오류가 발생했습니다"
            }
        }
    }

    fun leaveGroup() {
        val group = _selectedGroup.value ?: return
        val user = authService.getUserData() ?: return

        viewModelScope.launch {
            try {
                val mtIdx = user.mtIdx ?: return@launch
                val success = groupService.removeMember(group.sgtIdx, mtIdx)
                if (success) {
                    _successMessage.value = "그룹에서 탈퇴되었습니다"
                    hideDeleteDialog()
                    backToList()
                    loadGroups()
                    GlobalEventBus.tryEmit(GlobalEvent.GroupsChanged)
                } else {
                    _errorMessage.value = "그룹 탈퇴에 실패했습니다"
                }
            } catch (e: Exception) {
                _errorMessage.value = "그룹 탈퇴 중 오류가 발생했습니다"
            }
        }
    }

    fun updateMemberRole(isLeader: Boolean) {
        val group = _selectedGroup.value ?: return
        val member = _selectedMember.value ?: return

        viewModelScope.launch {
            try {
                val success = groupService.updateMemberRole(group.sgtIdx, member.mtIdx, isLeader)
                if (success) {
                    _successMessage.value = if (isLeader) "리더로 변경되었습니다" else "멤버로 변경되었습니다"
                    hideMemberManageDialog()
                    loadGroupMembers(group)
                } else {
                    _errorMessage.value = "역할 변경에 실패했습니다"
                }
            } catch (e: Exception) {
                _errorMessage.value = "역할 변경 중 오류가 발생했습니다"
            }
        }
    }

    fun removeMember() {
        val group = _selectedGroup.value ?: return
        val member = _selectedMember.value ?: return

        viewModelScope.launch {
            try {
                val success = groupService.removeMember(group.sgtIdx, member.mtIdx)
                if (success) {
                    _successMessage.value = "멤버가 탈퇴되었습니다"
                    hideMemberManageDialog()
                    loadGroupMembers(group)
                } else {
                    _errorMessage.value = "멤버 탈퇴에 실패했습니다"
                }
            } catch (e: Exception) {
                _errorMessage.value = "멤버 탈퇴 중 오류가 발생했습니다"
            }
        }
    }

    // MARK: - Helpers

    fun isCurrentUserOwner(): Boolean {
        val user = authService.getUserData() ?: return false
        val member = _groupMembers.value.find { it.mtIdx == user.mtIdx }
        return member?.sgdtOwnerChk == "Y"
    }

    fun getTotalMembers(): Int {
        return _groupMemberCounts.value.values.sum()
    }

    fun clearError() { _errorMessage.value = null }
    fun clearSuccess() { _successMessage.value = null }
}
